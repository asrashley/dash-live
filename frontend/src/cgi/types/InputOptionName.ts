export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}