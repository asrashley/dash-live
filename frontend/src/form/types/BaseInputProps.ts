import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

