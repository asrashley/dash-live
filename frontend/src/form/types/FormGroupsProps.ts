import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
