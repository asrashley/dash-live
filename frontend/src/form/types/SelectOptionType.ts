export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
