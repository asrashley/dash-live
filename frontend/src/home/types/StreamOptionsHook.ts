import { createContext } from "preact";
import { UseStreamOptionsHook } from "../hooks/useStreamOptions";

export const StreamOptionsContext = createContext<UseStreamOptionsHook>(null);
