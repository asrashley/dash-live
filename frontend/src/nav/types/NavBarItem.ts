export type NavBarItem = {
    className?: string;
    href: string;
    title: string;
};
