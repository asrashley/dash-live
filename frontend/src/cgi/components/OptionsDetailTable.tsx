import { Fragment } from "preact";
import { useComputed } from "@preact/signals";

import {
  useOptionsDetails,
  UseOptionsDetailsHook,
} from "../hooks/useOptionsDetails";
import { InputOptionName } from "../types/InputOptionName";
import { Icon } from "../../components/Icon";

function OptionsDetailRow({ shortName, fullName, cgiName }: InputOptionName) {
  return (
    <tr id={`opt_${shortName}`}>
      <td class="short-name">{shortName}</td>
      <td class="full-name">{fullName}</td>
      <td class="cgi-param">{cgiName}</td>
    </tr>
  );
}

interface SortableHeadingProps {
  name: string;
  setSortField: UseOptionsDetailsHook["setSortField"];
  sortAscending: UseOptionsDetailsHook["sortAscending"];
  sortField: UseOptionsDetailsHook["sortField"];
  title: string;
}

function SortableHeading({
  name,
  setSortField,
  sortField,
  sortAscending,
  title,
}: SortableHeadingProps) {
  const linkClass = `link${name === sortField.value ? " fw-bold" : ""}`;
  const iconClassName = name === sortField.value ? "opacity-100" : "opacity-50";
  const iconName = sortAscending.value ? "sort-alpha-down" : "sort-alpha-up";
  const onClick = () => {
    setSortField(name, name === sortField.value ? !sortAscending.value : true);
  };

  return (
    <a href="#" className={linkClass} onClick={onClick}>
      <span className="me-1">{title}</span>
      <Icon name={iconName} className={iconClassName} />
    </a>
  );
}

export function OptionsDetailTable() {
  const { allOptions, ...sortProps } = useOptionsDetails();
  const sortNameClass = useComputed<string>(
    () => `sort_${sortProps.sortField.value}`
  );

  return (
    <Fragment>
      <a id="json" />
      <table
        className="cgi-json-options table table-striped table-bordered"
        style="width: auto"
      >
        <thead>
          <tr className={sortNameClass}>
            <th className="short-name">
              <SortableHeading
                name="shortName"
                title="Short Name"
                {...sortProps}
              />
            </th>
            <th className="full-name">
              <SortableHeading
                name="fullName"
                title="Full Name"
                {...sortProps}
              />
            </th>
            <th className="cgi-param">
              <SortableHeading name="cgiName" title="CGI Name" {...sortProps} />
            </th>
          </tr>
        </thead>
        <tbody>
          {allOptions.value.map((opt) => (
            <OptionsDetailRow key={opt.shortName} {...opt} />
          ))}
        </tbody>
      </table>
    </Fragment>
  );
}
