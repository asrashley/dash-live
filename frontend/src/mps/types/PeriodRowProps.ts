import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
