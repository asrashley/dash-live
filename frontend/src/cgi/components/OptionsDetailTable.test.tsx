import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { signal } from "@preact/signals";
import { fireEvent } from "@testing-library/preact";

import { fieldGroups } from "@dashlive/options";

import { renderWithProviders } from "../../test/renderWithProviders";
import {
  type UseOptionsDetailsHook,
  useOptionsDetails,
} from "../hooks/useOptionsDetails";
import { InputOptionName } from "../types/InputOptionName";
import { OptionsDetailTable } from "./OptionsDetailTable";

vi.mock("../hooks/useOptionsDetails");

function createOptionNames(): InputOptionName[] {
  const names: InputOptionName[] = [];
  fieldGroups.forEach((grp) => {
    grp.fields.forEach(({ name, fullName, shortName }) => {
      names.push({
        cgiName: name,
        fullName,
        shortName,
      });
    });
  });
  return names;
}

describe("OptionsDetailTable component", () => {
  const allOptions = signal<InputOptionName[]>([]);
  const sortField = signal<string>("fullName");
  const sortAscending = signal<boolean>(true);
  const setSortField = vi.fn();
  const useOptionsHook: UseOptionsDetailsHook = {
    allOptions,
    sortField,
    sortAscending,
    setSortField,
  };
  const useOptionsDetailsMock = vi.mocked(useOptionsDetails);

  beforeEach(() => {
    useOptionsDetailsMock.mockReturnValue(useOptionsHook);
    allOptions.value = createOptionNames();
    sortField.value = "fullName";
    sortAscending.value = true;
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("matches snapshot", () => {
    const { asFragment, getBySelector } = renderWithProviders(
      <OptionsDetailTable />
    );
    fieldGroups.forEach((grp) => {
      grp.fields.forEach((field) => {
        const row = getBySelector(`#opt_${field.shortName}`) as HTMLElement;
        const full = row.querySelector(".full-name") as HTMLElement | null;
        expect(full).not.toBeNull();
        expect(full.innerHTML.trim()).toEqual(field.fullName);
        const sn = row.querySelector(".short-name") as HTMLElement | null;
        expect(sn).not.toBeNull();
        expect(sn.innerHTML.trim()).toEqual(field.shortName);
        const cgi = row.querySelector(".cgi-param") as HTMLElement | null;
        expect(cgi).not.toBeNull();
        expect(cgi.innerHTML.trim()).toEqual(field.name);
      });
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test.each([
    ["full-name", "fullName"],
    ["short-name", "shortName"],
    ["cgi-param", "cgiName"],
  ])("can sort by %s", (field: string, name: string) => {
    const { getBySelector } = renderWithProviders(<OptionsDetailTable />);
    const heading = getBySelector(`th.${field} > a`) as HTMLElement;
    fireEvent.click(heading);
    expect(setSortField).toHaveBeenCalledTimes(1);
    expect(setSortField).toHaveBeenCalledWith(name, name !== sortField.value);
  });
});
