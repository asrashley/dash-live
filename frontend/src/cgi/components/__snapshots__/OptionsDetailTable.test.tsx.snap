// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OptionsDetailTable component > matches snapshot 1`] = `
<DocumentFragment>
  <a
    id="json"
  />
  <table
    class="cgi-json-options table table-striped table-bordered"
    style="width: auto;"
  >
    <thead>
      <tr
        class="sort_fullName"
      >
        <th
          class="short-name"
        >
          <a
            class="link"
            href="#"
          >
            <span
              class="me-1"
            >
              Short Name
            </span>
            <span
              aria-label="sort-alpha-down icon"
              class="bi icon bi-sort-alpha-down opacity-50"
              role="img"
            />
          </a>
        </th>
        <th
          class="full-name"
        >
          <a
            class="link fw-bold"
            href="#"
          >
            <span
              class="me-1"
            >
              Full Name
            </span>
            <span
              aria-label="sort-alpha-down icon"
              class="bi icon bi-sort-alpha-down opacity-100"
              role="img"
            />
          </a>
        </th>
        <th
          class="cgi-param"
        >
          <a
            class="link"
            href="#"
          >
            <span
              class="me-1"
            >
              CGI Name
            </span>
            <span
              aria-label="sort-alpha-down icon"
              class="bi icon bi-sort-alpha-down opacity-50"
              role="img"
            />
          </a>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        id="opt_ab"
      >
        <td
          class="short-name"
        >
          ab
        </td>
        <td
          class="full-name"
        >
          abr
        </td>
        <td
          class="cgi-param"
        >
          abr
        </td>
      </tr>
      <tr
        id="opt_ac"
      >
        <td
          class="short-name"
        >
          ac
        </td>
        <td
          class="full-name"
        >
          audioCodec
        </td>
        <td
          class="cgi-param"
        >
          acodec
        </td>
      </tr>
      <tr
        id="opt_evs"
      >
        <td
          class="short-name"
        >
          evs
        </td>
        <td
          class="full-name"
        >
          eventTypes
        </td>
        <td
          class="cgi-param"
        >
          events
        </td>
      </tr>
      <tr
        id="opt_st"
      >
        <td
          class="short-name"
        >
          st
        </td>
        <td
          class="full-name"
        >
          segmentTimeline
        </td>
        <td
          class="cgi-param"
        >
          timeline
        </td>
      </tr>
      <tr
        id="opt_utc"
      >
        <td
          class="short-name"
        >
          utc
        </td>
        <td
          class="full-name"
        >
          utcMethod
        </td>
        <td
          class="cgi-param"
        >
          time
        </td>
      </tr>
      <tr
        id="opt_vp"
      >
        <td
          class="short-name"
        >
          vp
        </td>
        <td
          class="full-name"
        >
          videoPlayer
        </td>
        <td
          class="cgi-param"
        >
          player
        </td>
      </tr>
      <tr
        id="opt_ad"
      >
        <td
          class="short-name"
        >
          ad
        </td>
        <td
          class="full-name"
        >
          audioDescription
        </td>
        <td
          class="cgi-param"
        >
          ad_audio
        </td>
      </tr>
      <tr
        id="opt_ast"
      >
        <td
          class="short-name"
        >
          ast
        </td>
        <td
          class="full-name"
        >
          availabilityStartTime
        </td>
        <td
          class="cgi-param"
        >
          start
        </td>
      </tr>
      <tr
        id="opt_bug"
      >
        <td
          class="short-name"
        >
          bug
        </td>
        <td
          class="full-name"
        >
          bugCompatibility
        </td>
        <td
          class="cgi-param"
        >
          bugs
        </td>
      </tr>
      <tr
        id="opt_dft"
      >
        <td
          class="short-name"
        >
          dft
        </td>
        <td
          class="full-name"
        >
          clockDrift
        </td>
        <td
          class="cgi-param"
        >
          drift
        </td>
      </tr>
      <tr
        id="opt_lee"
      >
        <td
          class="short-name"
        >
          lee
        </td>
        <td
          class="full-name"
        >
          leeway
        </td>
        <td
          class="cgi-param"
        >
          leeway
        </td>
      </tr>
      <tr
        id="opt_ma"
      >
        <td
          class="short-name"
        >
          ma
        </td>
        <td
          class="full-name"
        >
          mainAudio
        </td>
        <td
          class="cgi-param"
        >
          main_audio
        </td>
      </tr>
      <tr
        id="opt_mt"
      >
        <td
          class="short-name"
        >
          mt
        </td>
        <td
          class="full-name"
        >
          mainText
        </td>
        <td
          class="cgi-param"
        >
          main_text
        </td>
      </tr>
      <tr
        id="opt_mup"
      >
        <td
          class="short-name"
        >
          mup
        </td>
        <td
          class="full-name"
        >
          minimumUpdatePeriod
        </td>
        <td
          class="cgi-param"
        >
          mup
        </td>
      </tr>
      <tr
        id="opt_ntps"
      >
        <td
          class="short-name"
        >
          ntps
        </td>
        <td
          class="full-name"
        >
          ntpSources
        </td>
        <td
          class="cgi-param"
        >
          ntp_servers
        </td>
      </tr>
      <tr
        id="opt_tc"
      >
        <td
          class="short-name"
        >
          tc
        </td>
        <td
          class="full-name"
        >
          textCodec
        </td>
        <td
          class="cgi-param"
        >
          tcodec
        </td>
      </tr>
      <tr
        id="opt_tl"
      >
        <td
          class="short-name"
        >
          tl
        </td>
        <td
          class="full-name"
        >
          textLanguage
        </td>
        <td
          class="cgi-param"
        >
          tlang
        </td>
      </tr>
      <tr
        id="opt_base"
      >
        <td
          class="short-name"
        >
          base
        </td>
        <td
          class="full-name"
        >
          useBaseUrls
        </td>
        <td
          class="cgi-param"
        >
          base
        </td>
      </tr>
      <tr
        id="opt_patch"
      >
        <td
          class="short-name"
        >
          patch
        </td>
        <td
          class="full-name"
        >
          patch
        </td>
        <td
          class="cgi-param"
        >
          patch
        </td>
      </tr>
      <tr
        id="opt_tbd"
      >
        <td
          class="short-name"
        >
          tbd
        </td>
        <td
          class="full-name"
        >
          timeShiftBufferDepth
        </td>
        <td
          class="cgi-param"
        >
          depth
        </td>
      </tr>
      <tr
        id="opt_clearkey__enabled"
      >
        <td
          class="short-name"
        >
          clearkey__enabled
        </td>
        <td
          class="full-name"
        >
          enabled
        </td>
        <td
          class="cgi-param"
        >
          clearkey__enabled
        </td>
      </tr>
      <tr
        id="opt_clu"
      >
        <td
          class="short-name"
        >
          clu
        </td>
        <td
          class="full-name"
        >
          licenseUrl
        </td>
        <td
          class="cgi-param"
        >
          clearkey__la_url
        </td>
      </tr>
      <tr
        id="opt_clearkey__drmloc"
      >
        <td
          class="short-name"
        >
          clearkey__drmloc
        </td>
        <td
          class="full-name"
        >
          drmloc
        </td>
        <td
          class="cgi-param"
        >
          clearkey__drmloc
        </td>
      </tr>
      <tr
        id="opt_marlin__enabled"
      >
        <td
          class="short-name"
        >
          marlin__enabled
        </td>
        <td
          class="full-name"
        >
          enabled
        </td>
        <td
          class="cgi-param"
        >
          marlin__enabled
        </td>
      </tr>
      <tr
        id="opt_mlu"
      >
        <td
          class="short-name"
        >
          mlu
        </td>
        <td
          class="full-name"
        >
          licenseUrl
        </td>
        <td
          class="cgi-param"
        >
          marlin__la_url
        </td>
      </tr>
      <tr
        id="opt_marlin__drmloc"
      >
        <td
          class="short-name"
        >
          marlin__drmloc
        </td>
        <td
          class="full-name"
        >
          drmloc
        </td>
        <td
          class="cgi-param"
        >
          marlin__drmloc
        </td>
      </tr>
      <tr
        id="opt_pinCoun"
      >
        <td
          class="short-name"
        >
          pinCoun
        </td>
        <td
          class="full-name"
        >
          count
        </td>
        <td
          class="cgi-param"
        >
          ping__count
        </td>
      </tr>
      <tr
        id="opt_pinDura"
      >
        <td
          class="short-name"
        >
          pinDura
        </td>
        <td
          class="full-name"
        >
          duration
        </td>
        <td
          class="cgi-param"
        >
          ping__duration
        </td>
      </tr>
      <tr
        id="opt_pinInba"
      >
        <td
          class="short-name"
        >
          pinInba
        </td>
        <td
          class="full-name"
        >
          inband
        </td>
        <td
          class="cgi-param"
        >
          ping__inband
        </td>
      </tr>
      <tr
        id="opt_pinInte"
      >
        <td
          class="short-name"
        >
          pinInte
        </td>
        <td
          class="full-name"
        >
          interval
        </td>
        <td
          class="cgi-param"
        >
          ping__interval
        </td>
      </tr>
      <tr
        id="opt_pinStar"
      >
        <td
          class="short-name"
        >
          pinStar
        </td>
        <td
          class="full-name"
        >
          start
        </td>
        <td
          class="cgi-param"
        >
          ping__start
        </td>
      </tr>
      <tr
        id="opt_pinTime"
      >
        <td
          class="short-name"
        >
          pinTime
        </td>
        <td
          class="full-name"
        >
          timescale
        </td>
        <td
          class="cgi-param"
        >
          ping__timescale
        </td>
      </tr>
      <tr
        id="opt_pinValu"
      >
        <td
          class="short-name"
        >
          pinValu
        </td>
        <td
          class="full-name"
        >
          value
        </td>
        <td
          class="cgi-param"
        >
          ping__value
        </td>
      </tr>
      <tr
        id="opt_pinVers"
      >
        <td
          class="short-name"
        >
          pinVers
        </td>
        <td
          class="full-name"
        >
          version
        </td>
        <td
          class="cgi-param"
        >
          ping__version
        </td>
      </tr>
      <tr
        id="opt_playready__enabled"
      >
        <td
          class="short-name"
        >
          playready__enabled
        </td>
        <td
          class="full-name"
        >
          enabled
        </td>
        <td
          class="cgi-param"
        >
          playready__enabled
        </td>
      </tr>
      <tr
        id="opt_plu"
      >
        <td
          class="short-name"
        >
          plu
        </td>
        <td
          class="full-name"
        >
          licenseUrl
        </td>
        <td
          class="cgi-param"
        >
          playready__la_url
        </td>
      </tr>
      <tr
        id="opt_pff"
      >
        <td
          class="short-name"
        >
          pff
        </td>
        <td
          class="full-name"
        >
          piff
        </td>
        <td
          class="cgi-param"
        >
          playready__piff
        </td>
      </tr>
      <tr
        id="opt_pvn"
      >
        <td
          class="short-name"
        >
          pvn
        </td>
        <td
          class="full-name"
        >
          version
        </td>
        <td
          class="cgi-param"
        >
          playready__version
        </td>
      </tr>
      <tr
        id="opt_playready__drmloc"
      >
        <td
          class="short-name"
        >
          playready__drmloc
        </td>
        <td
          class="full-name"
        >
          drmloc
        </td>
        <td
          class="cgi-param"
        >
          playready__drmloc
        </td>
      </tr>
      <tr
        id="opt_sctCoun"
      >
        <td
          class="short-name"
        >
          sctCoun
        </td>
        <td
          class="full-name"
        >
          count
        </td>
        <td
          class="cgi-param"
        >
          scte35__count
        </td>
      </tr>
      <tr
        id="opt_sctDura"
      >
        <td
          class="short-name"
        >
          sctDura
        </td>
        <td
          class="full-name"
        >
          duration
        </td>
        <td
          class="cgi-param"
        >
          scte35__duration
        </td>
      </tr>
      <tr
        id="opt_sctInba"
      >
        <td
          class="short-name"
        >
          sctInba
        </td>
        <td
          class="full-name"
        >
          inband
        </td>
        <td
          class="cgi-param"
        >
          scte35__inband
        </td>
      </tr>
      <tr
        id="opt_sctInte"
      >
        <td
          class="short-name"
        >
          sctInte
        </td>
        <td
          class="full-name"
        >
          interval
        </td>
        <td
          class="cgi-param"
        >
          scte35__interval
        </td>
      </tr>
      <tr
        id="opt_sctProg"
      >
        <td
          class="short-name"
        >
          sctProg
        </td>
        <td
          class="full-name"
        >
          program_id
        </td>
        <td
          class="cgi-param"
        >
          scte35__program_id
        </td>
      </tr>
      <tr
        id="opt_sctStar"
      >
        <td
          class="short-name"
        >
          sctStar
        </td>
        <td
          class="full-name"
        >
          start
        </td>
        <td
          class="cgi-param"
        >
          scte35__start
        </td>
      </tr>
      <tr
        id="opt_sctTime"
      >
        <td
          class="short-name"
        >
          sctTime
        </td>
        <td
          class="full-name"
        >
          timescale
        </td>
        <td
          class="cgi-param"
        >
          scte35__timescale
        </td>
      </tr>
      <tr
        id="opt_sctValu"
      >
        <td
          class="short-name"
        >
          sctValu
        </td>
        <td
          class="full-name"
        >
          value
        </td>
        <td
          class="cgi-param"
        >
          scte35__value
        </td>
      </tr>
      <tr
        id="opt_sctVers"
      >
        <td
          class="short-name"
        >
          sctVers
        </td>
        <td
          class="full-name"
        >
          version
        </td>
        <td
          class="cgi-param"
        >
          scte35__version
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
